### ISSUE TITLE
Game will not start after update 12-6-20 (v2.0.0-alpha1)

### ISSUE THREAD

**Issue by Tater-K7S5A** [lexica/lexica#211]
I am running Android 11 on a Pixel 2XL.  This morning I had an update for Lexica and ran it.  After updating, I am unable to start the game in the application.

The application launches and I am able to access the menus for the various settings.  If I select "New Game" the screen blanks and returns to the initial page with "New Game", "About", "Restore Game", "Preferences".  If I persist in trying to start a game I get an Android system message telling me "Lexica keeps stopping" and gives me "App Info" and "Close app" options.  Selecting "Restore Game" has no effect or sign of actually doing anything.  I can select it as many times as I want and get no Android system message.

I have restarted the phone with no change to the behavior.   

I uninstalled Lexica and reinstalled it with no change to the behavior.

**Comment by NullDepot** [comment]
I've just installed the app for the first time after discovering it on F-droid and can't start a new game either. Same issue as mentioned above. I'm running GrapheneOS on a Pixel 4XL.

**Comment by msteudtn** [comment]
I can confirm the problem. Lexica crashes after selecting "New Game". :( 

* LineageOS 14.1 (Android 7.1) 
* F-Droid version - 2.0.0-alpha1
* no GAPPS

**Comment by darhma** [comment]
Same issue with F-Droid version 2.0.0-alpha1 running on Xiaomi Redmi Note 7 with lineageos 17.1

**Comment by pserwylo** [comment]
Oh gosh, thanks so much for the quick feedback everyone. I can confirm that this is related to the following crash:

```
         AndroidRuntime  E  FATAL EXCEPTION: main
                         E  Process: com.serwylo.lexica, PID: 18155
                         E  java.lang.RuntimeException: Unable to resume activity {com.serwylo.lexica/com.serwylo.lexica.GameActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'com.serwy
                            lo.lexica.game.Game$GameStatus com.serwylo.lexica.game.Game.getStatus()' on a null object reference
                         E      at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4902)
                         E      at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4945)
                         E      at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
                         E      at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:190)
                         E      at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:105)
                         E      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2473)
                         E      at android.os.Handler.dispatchMessage(Handler.java:110)
                         E      at android.os.Looper.loop(Looper.java:219)
                         E      at android.app.ActivityThread.main(ActivityThread.java:8347)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1055)
                         E  Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'com.serwylo.lexica.game.Game$GameStatus com.serwylo.lexica.game.Game.getStatus()' on a null object referen
                            ce
                         E      at com.serwylo.lexica.GameActivity.onResume(GameActivity.java:206)
                         E      at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1472)
                         E      at android.app.Activity.performResume(Activity.java:8222)
                         E      at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4892)
                         E      ... 11 more
```

I'll see if I can get quick fix out and release an `-alpha2`.

I also realise it is bad that you have all likely updated because F-Droid told you to, even though it is an Alpha release. That is also my fault, I've incorrectly misconfigured the F-Droid metadata such that all versions become the "suggested version". I'll leave it like that until this bug is fixed (so that new users and others updating don't get this bad bug), but after that, I'll try to fix the metadata so alpha builds are not thrust upon you!

**Comment by pserwylo** [comment]
Initial investigation shows that a local debug build from `git` works as expected, however the F-Droid build fails with the following earlier error:

```
                         E  android.content.res.Resources$NotFoundException: Resource ID #0x0
                         E      at android.content.res.ResourcesImpl.getValue(ResourcesImpl.java:328)
                         E      at android.content.res.ResourcesImpl.openRawResource(ResourcesImpl.java:467)
                         E      at android.content.res.Resources.openRawResource(Resources.java:1418)
                         E      at android.content.res.Resources.openRawResource(Resources.java:1362)
                         E      at com.serwylo.lexica.game.Game.initializeDictionary(Game.java:270)
                         E      at com.serwylo.lexica.game.Game.initializeDictionary(Game.java:263)
                         E      at com.serwylo.lexica.game.Game.setBoard(Game.java:235)
                         E      at com.serwylo.lexica.game.Game.<init>(Game.java:184)
                         E      at com.serwylo.lexica.GameActivity.newGame(GameActivity.java:124)
                         E      at com.serwylo.lexica.GameActivity.onCreate(GameActivity.java:91)
                         E      at android.app.Activity.performCreate(Activity.java:8085)
                         E      at android.app.Activity.performCreate(Activity.java:8073)
                         E      at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1320)
                         E      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3868)
                         E      at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4074)
                         E      at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:91)
                         E      at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:149)
                         E      at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:103)
                         E      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2473)
                         E      at android.os.Handler.dispatchMessage(Handler.java:110)
                         E      at android.os.Looper.loop(Looper.java:219)
                         E      at android.app.ActivityThread.main(ActivityThread.java:8347)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1055)
```

This happens during `onCreate`, and implies that the release build doesn't end up with `res/raw/dictionary_LANG` files :(

I will have to look into why these are not present.

The good news is that it si reproducible locally if I run the F-Droid build here. It also means that I should add to my test plan 'Quick regression test on the F-Droid build prior to tagging for release".

**Comment by pserwylo** [comment]
I haven't yet been able to figure out why these resources are not included in the resulting `.apk` from `fdroidserver`. 

Things I've tried so far:
* Confirmed that `shrinkResources` is `false` in Lexica (the dictionaries are referenced dynamically in code, so I was concerned some sort of Gradle optimisation was stripping them out).
* The last version of Lexica successfully built was v1.3.1
** https://f-droid.org/wiki/page/com.serwylo.lexica/lastbuild_10301
** fdroidserverid: [1f7228d53859f078698468b2e0bbe48485e4ece7](https://gitlab.com/fdroid/fdroidserver/commit/1f7228d53859f078698468b2e0bbe48485e4ece7)
* The failed build v2.0.0-alpha1 used a different version of F-Droid server
** https://f-droid.org/wiki/page/com.serwylo.lexica/lastbuild_20000
** fdroidserverid: [767573f2ecf3674a31d6ca4c7c7e05c4ea52f07e](https://gitlab.com/fdroid/fdroidserver/commit/767573f2ecf3674a31d6ca4c7c7e05c4ea52f07e)
* I've now tried running `fdroid build` with these two versions of `fdroidserver`, but I don't yet have the correct permutations:
** v1.3.1 with old `fdroidserver` (should work)
** v1.3.1 with new `fdroidserver` (if it works, then start debugging what has changed between two versions of `fdroidserver`)
** v2.0.0-alpha1 with new `fdroidserver` (should fail)
** v2.0.0-alpha1 with old `fdroidserver` (if it fails, then start debugging what has changed between two versions of Lexica - including the introduction of Kotlin which had the biggest impact on build scripts).

I will keep investigating when I get some time over the coming days. Thanks for your patience.

**Comment by pserwylo** [comment]
Furthermore, the existing version v1.3.1 uses Gradle 6.1.1 whereas the new version uses v2.0.0.

**Comment by Tater-K7S5A** [comment]
Thank you for your efforts.  I really appreciate it.

**Comment by pserwylo** [comment]
Okay, so I think I have a fix, though I still don't understand the root cause.

We have a task to build the dictionary files for use by Lexica. This `gradle` tasks is marked such that the `assembleRelease` task depends on it:

```
afterEvaluate {
  assembleDebug.dependsOn rootProject.buildDictionaries
  if (tasks.findByPath('assembleRelease')) {
    assembleRelease.dependsOn rootProject.buildDictionaries
  }
}
```

I was under the assumption this would be deterministic, in that in order to "assemble a release", we always need the dictionaries to be present. This was incorrect. All this does is put the `buildDictionaries` task at the same level as _all the other_ dependencies of this. As such, when run locally, it seemed to always run some time nice and early prior to Gradle gathering the resource for the apk. However, when run by `fdroidserver`, it _always_ made this happen _just before_ the `assembleRelease` task (but after we had done all of the actual assembling).

What I couldn't figure out is _why_ this is the case, but it is likely something to do with the fact `fdroidserver` has a requirement to have its own `fdroid-gradlew`.

Anyway, turns out I can be more explicit, and make the dependencies a little better:

```
afterEvaluate {
   generateDebugResources.dependsOn rootProject.buildDictionaries
   generateDebugResources.shouldRunAfter rootProject.buildDictionaries
   if (tasks.findByPath('generateReleaseResources')) {
       generateReleaseResources.dependsOn rootProject.buildDictionaries
       generateReleaseResources.shouldRunAfter rootProject.buildDictionaries
    }
}
```

This not only makes the `buildDictionaries` actually a requirement to happen before the `generate*Resources` stage (earlier on than just `assemble*Release`), but it also forces that if for whatever reason both tasks are specified on the same gradle command, that the `buildDictionaries` task will happen first.

Thanks for your reports and patience, I've tagged `v2.0.0-alpha2` and once that is picked up by F-Droid, we can close this if we are able to show that it works as expected.

**Comment by pserwylo** [comment]
Closed via merge request, but want to keep open until made available in F-Droid.

**Comment by pserwylo** [comment]
Phew, v2.0.0-alpha2 is now out on F-Droid, and it seems to work (at least, this bug is resolved). Thanks all. Feedback on the new version is of course welcome via the issue tracker too.

**Comment by rugk** [comment]
Also related: https://github.com/lexica/lexica/issues/214

**Comment by Tater-K7S5A** [comment]
Thank you for your hard work.  I got the update notification this morning and the game appears to be working correctly.

**Comment by msteudtn** [comment]
:+1: works for me again, too. :-)

**Comment by glha** [comment]
Alpha2 is live now and functional.

**Comment by Eaglechild** [comment]
I updated Lexica over F-droid today and the game does not start. The screen flickers once, then I see the apps tray again. No message, no crash report. I am on /e/ OS (Android 8) without Google Apps. The previous version was functioning, so I installed the apk of Version 1.3.1 (10301) from F-droid again. But the Version 2.0.0-alpha1 (20000) will not work at all on my phone.
I love this game! I play it whenever I need to wait a little, in public transport etc. Is there a chance that I can play the new version?
Thanks a lot for your work!

**Comment by darhma** [comment]
@Eaglechild you have to install version 2.0.0-alpha2 that solves the issue. You can find it on f-droid (maybe you have to update the repository)

**Comment by Eaglechild** [comment]
Thanks for your answer, darhma, but still no luck, I'm afraid. I installed version 2.0.0-alpha2 with the f-droid app as it is the update the app suggests for Lexica. But the behavior is the same: screen flickers once, then I see the apps tray. When I go to the f-droid side with my PC, I can't find this version. The version that is suggested there is 2.0.0-alpha1. How can I update the f-droid repository? I could not find any link on the website to do this.
Would it help to delete my Lexica data from the previous version? But I would not know how to do that either...
Thanks in advance for your answer.

**Comment by darhma** [comment]
@Eaglechild you can go to the Lexica's app info in the device settings and select "storage & cache" and then "clear storage" to delete the data of previous versions. Or you can try to uninstall previous versions of Lexica and then install directly version 2.0.0-alpha2 from f-droid app (this should delete the data of previous versions).

**Comment by Eaglechild** [comment]
thanks for your answer, darhma.
So I deleted all data with your instructions, then I uninstalled Lexica. I uninstalled and reinstalled the f-droid app, and it updated the repository. I installed version 2.0.0-alpha2 of Lexica from the updated repository, but the game still won't start. I even tried the same with another phone with stock Android 8, where Lexica has never been installed before, but there it does not start either.
Something is still missing here.
Strangely, when I go to the f-droid website with my computer running Linux, version 2.0.0-alpha1 shows up as the latest version. So people downloading the latest apk will all have the bug from that version. Also I can't try to install the alpha2 version from the apk.
I will try to install the latest version on a third phone with stock Android 10 in the afternoon and post the results.
